backends/platform/3ds/main.cpp

```diff
@@ -25,16 +25,27 @@
 #include <3ds.h>
 #include <malloc.h>
 
+// 0x1FF80030 holds the address of a u32 value representing which of several
+//	possible FCRAM layouts the hardware kernel spports. Layouts 0 through 5
+//	are used only by Old 3DS kernels, while layouts 6 through 8 are only used
+//	by New 3DS kernels.
+#define APPMEMTYPE (*(u32*)0x1FF80030)
+
 enum {
 	SYSTEM_MODEL_2DS = 3
 };
 
 // Set the size of the stack.
 u32 __stacksize__ = 64 * 1024;
 
+// New 3DS needs a larger linear heap to boot the app than the Old 3DS does.
+// The reason we are setting the memory size conditionally based on the system model,
+//	instead of just using 20MB, is to allow for future functionality coming soon(TM).
+u32 mem = (APPMEMTYPE < 6) ? 10 : 20;
```

Comment:
I got it from [here](https://github.com/TuxSH/universal-otherapp/blob/f6c67f520c8e38cf6a1ab7c68600dd4af827ed3c/source/lib/defines.h#L13), but it is also used in [many other 3DS homebrew apps](https://github.com/search?q=0x1FF80030&type=code). TuxSH (the code author in the first link) could probably explain it better than me, since they're the lead developer of the [Luma3DS CFW (custom firmware)](https://github.com/LumaTeam/Luma3DS).

Other documentation:
[3Dbrew: Configuration_Memory - ARM11 Kernel Configuration Memory](https://www.3dbrew.org/wiki/Configuration_Memory#ARM11_Kernel_Configuration_Memory)
[3Dbrew: Configuration_Memory - APPMEMTYPE](https://www.3dbrew.org/wiki/Configuration_Memory#APPMEMTYPE)
[3Dbrew: Memory layout - FCRAM memory-regions layout](https://www.3dbrew.org/wiki/Memory_layout#FCRAM_memory-regions_layout)

Regardless, it DOES work.

Reply:
So, I did some tests.
First, I don't question the use of `APPMEMTYPE` which is a macro dereferencing some memory location which contains a value as specified in some datasheet. This define is perfectly legit.
What I question is its use in a global variable which is expected to be set at compilation time.
This is not possible as the `APPMEMTYPE` is not known at this time.
What I didn't have in sight is that, in C++, doing such global initialization with non constant values is legal, while it's not in C.
In C++, the compiler will resort to generate some piece of code to set up the value and make it call at initialization time.
Now... when is the initialization? This is done by the `__libc_init_array` function, called in the [`initSystem`](https://github.com/devkitPro/libctru/blob/8cbcce773b957684577b3a77c42d2c2864b016af/libctru/source/system/stack_adjust.s#L21) function of libctru.
But when is the `__ctru_linear_heap_size` used? It's first use is seems to be in [`__system_allocateHeaps`](https://github.com/devkitPro/libctru/blob/8cbcce773b957684577b3a77c42d2c2864b016af/libctru/source/system/allocateHeaps.c#L38) which is called by `__libctru_init`, which is called in `initSystem` before `__libc_init_array`.
So, what you have at this time, is a `__ctru_linear_heap_size` variable (located in `.bss`) initialized to 0.
This means (from libctru source code):
> Split available memory equally between linear and application heaps (with rounding in favor of the latter).

So, you get more of linear size, as you need for ND3S, but also less heap size, which wasn't intended.

The links you provided are correct but they never use this define as you do: globally.
They only use the define in some function code where it's working as intended.

What could be done, as `__system_allocateHeaps` is weak defined, is to rewrite it in our code to do what we want.
Another way could be to use the `--wrap` option of `ld` to wrap the ctru function around some code which will set the variable accordingly before calling the real one.

Reply:
You're right. I inserted some code to print `envGetHeapSize()` and `envGetLinearHeapSize()` to `scummvm.log`; just straight changing the linear mem size to 20MiB gives the heap 96632832 bytes (~92.16 MiB), while my way only gives the heap 84049920 bytes (~80.16 MiB). I'll go back to the drawing board, and I might take you up on overriding `__system_AllocateHeaps`.